wfe/web-front-end.go

```diff
@@ -1186,6 +1202,68 @@ func (wfe *WebFrontEndImpl) BuildID(response http.ResponseWriter, request *http.
 	}
 }
 
+// Options responds to an HTTP OPTIONS request.
+func (wfe *WebFrontEndImpl) Options(response http.ResponseWriter, request *http.Request, methodsStr string, methodsMap map[string]bool) {
+	// Every OPTIONS request gets an Allow header with a list of supported methods.
+	response.Header().Set("Allow", methodsStr)
+
+	// CORS preflight requests get additional headers. See
+	// http://www.w3.org/TR/cors/#resource-preflight-requests
+	reqMethod := request.Header.Get("Access-Control-Request-Method")
+	if reqMethod == "" {
+		reqMethod = "GET"
+	}
+	if methodsMap[reqMethod] {
+		wfe.setCORSHeaders(response, request, methodsStr)
+	}
+}
+
+// setCORSHeaders() tells the client that CORS is acceptable for this
+// request. If allowMethods == "" the request is assumed to be a CORS
+// actual request and no Access-Control-Allow-Methods or -Headers
+// headers will be sent.
+func (wfe *WebFrontEndImpl) setCORSHeaders(response http.ResponseWriter, request *http.Request, allowMethods string) {
+	reqOrigin := request.Header.Get("Origin")
+	if reqOrigin == "" {
+		// This is not a CORS request.
+		return
+	}
+
+	// Allow CORS if the current origin (or "*") is listed as an
+	// allowed origin in config. Otherwise, disallow by returning
+	// without setting any CORS headers.
+	allow := false
+	for _, ao := range wfe.AllowOrigins {
+		if ao == "*" {
+			response.Header().Set("Access-Control-Allow-Origin", "*")
+			allow = true
+			break
+		} else if ao == reqOrigin {
+			response.Header().Set("Vary", "Origin")
+			response.Header().Set("Access-Control-Allow-Origin", ao)
+			allow = true
+			break
+		}
+	}
+	if !allow {
+		return
+	}
+
+	if allowMethods != "" {
+		// For an OPTIONS request: allow all methods handled at this URL.
+		response.Header().Set("Access-Control-Allow-Methods", allowMethods)
+
+		// Allow all requested headers.
+		if acrh, ok := request.Header["Access-Control-Request-Headers"]; ok {
+			for _, h := range acrh {
+				response.Header().Add("Access-Control-Allow-Headers", h)
```

Comment:
I don't know of any reason a client would need to send a non-simple request header. I appreciate the thoroughness in implementing the CORS spec, but lets leave this part of it out until and unless we introduce non-simple request headers that are required by ACME.